### PR TITLE
fix: Add support for Python 3.11

### LINKED ISSUE #326: Test failures in Python 3.11.0b3
#### Environment details

  - Programming language: `Python`
  - OS: `Fedora 37 (rawhide)`
  - Language runtime version: `Python 3.11.0b3`
  - Package version: `v1.20.6`

#### Steps to reproduce

  1. Run `pytest` in Python 3.11.0b3

The first failure appears in `test_total_ordering_w_other_enum_type`:
```
____________________ test_total_ordering_w_other_enum_type _____________________
    def test_total_ordering_w_other_enum_type():
        to_compare = enums_test.OneEnum.SOME_VALUE
    
        for item in enums_test.OtherEnum:
            assert not to_compare == item
>           assert to_compare.SOME_VALUE != item
tests/test_enum_total_ordering.py:52: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = <enum.property object at 0x3ffa350e550>
instance = <OneEnum.SOME_VALUE: 1>, ownerclass = <enum 'OneEnum'>
    def __get__(self, instance, ownerclass=None):
        if instance is None:
            try:
                return ownerclass._member_map_[self.name]
            except KeyError:
                raise AttributeError(
                        '%r has no attribute %r' % (ownerclass, self.name)
                        )
        else:
            if self.fget is None:
>               raise AttributeError(
                        '%r member has no attribute %r' % (ownerclass, self.name)
                        )
E               AttributeError: <enum 'OneEnum'> member has no attribute 'SOME_VALUE'
/usr/lib64/python3.11/enum.py:198: AttributeError
```

The second is in `test_enum_alias_good`:
```
_____________________________ test_enum_alias_good _____________________________
    def test_enum_alias_good():
        # Have to split good and bad enum alias into two tests so that the generated
        # file descriptor is properly created.
        # For the python runtime, aliases are allowed by default, but we want to
        # make sure that the options don't cause problems.
        # For the cpp runtime, we need to verify that we can in fact define aliases.
>       class GoodMessage(proto.Message):
tests/test_fields_enum.py:386: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/test_fields_enum.py:387: in GoodMessage
    class GoodEnum(proto.Enum):
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
mcls = <class 'proto.enums.ProtoEnumMeta'>, name = 'GoodEnum'
bases = (<enum 'Enum'>,)
attrs = {'_generate_next_value_': <function Enum._generate_next_value_ at 0x3ffa5a8f560>, '__module__': 'test_fields_enum', '__qualname__': 'test_enum_alias_good.<locals>.GoodMessage.GoodEnum', 'UNKNOWN': 0, 'DEFAULT': 0}
    def __new__(mcls, name, bases, attrs):
        # Do not do any special behavior for `proto.Enum` itself.
        if bases[0] == enum.IntEnum:
            return super().__new__(mcls, name, bases, attrs)
    
        # Get the essential information about the proto package, and where
        # this component belongs within the file.
        package, marshal = _package_info.compile(name, attrs)
    
        # Determine the local path of this proto component within the file.
        local_path = tuple(attrs.get("__qualname__", name).split("."))
    
        # Sanity check: We get the wrong full name if a class is declared
        # inside a function local scope; correct this.
        if "<locals>" in local_path:
            ix = local_path.index("<locals>")
            local_path = local_path[: ix - 1] + local_path[ix + 1 :]
    
        # Determine the full name in protocol buffers.
        full_name = ".".join((package,) + local_path).lstrip(".")
        filename = _file_info._FileInfo.proto_file_name(
            attrs.get("__module__", name.lower())
        )
    
        # Retrieve any enum options.
        # We expect something that looks like an EnumOptions message,
        # either an actual instance or a dict-like representation.
        pb_options = "_pb_options"
        opts = attrs.pop(pb_options, {})
        # This is the only portable way to remove the _pb_options name
        # from the enum attrs.
        # In 3.7 onwards, we can define an _ignore_ attribute and do some
        # mucking around with that.
        if pb_options in attrs._member_names:
>           idx = attrs._member_names.index(pb_options)
E           AttributeError: 'dict' object has no attribute 'index'
../../BUILDROOT/python-proto-plus-1.20.6-1.fc37.noarch/usr/lib/python3.11/site-packages/proto/enums.py:61: AttributeError
```

For the second one, one of my coworkers found that `_member_names` is now a `dict` from a [recent PR](https://github.com/python/cpython/pull/28907).

### PR DESCRIPTION
**Pull request by hroncok**
There are two changes:

Changes in the actual code:

 - _member_names changed from a list to a dict in https://github.com/python/cpython/pull/28907
    - we instance-check and remove by list-specific or dict-specific way

Change in the tests only:

 - accessing other enum members via instance attributes is no longer possible
    - we access them via the class instead
    - we leave the original test in a try-except block

Some of the Python enum changes might get reverted,
see https://github.com/python/cpython/issues/93910
But the fix is backwards compatible.

Fixes https://github.com/googleapis/proto-plus-python/issues/326

### CODE REVIEW

**Comment by hroncok** [comment]
> Some of your commit messages failed linting.

Feel free to reword that message however you see fit.

> Thanks for your pull request! It looks like this may be your first contribution to a Google open source project. Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).

A link would be helpful for this.

File: tests/test_enum_total_ordering.py [hunk 1]
```diff
@@ -49,7 +49,11 @@ def test_total_ordering_w_other_enum_type():
 
     for item in enums_test.OtherEnum:
         assert not to_compare == item
-        assert to_compare.SOME_VALUE != item
+        assert type(to_compare).SOME_VALUE != item
```
Comment (ethanfurman) [inline tests/test_enum_total_ordering.py:52]: `to_compare` is already `OneEnum.SOME_VALUE` -- so what is the purpose of the indirect reference through itself?  Can this part of the test just be dropped?

**Comment by parthea** [comment]
Thanks @hroncok! Please could you sign the CLA using the link below?

-------

📝 If you are not currently covered under a CLA, please visit https://cla.developers.google.com/.

**Comment by parthea** [comment]
The github actions checks appear to be stuck with `Waiting for status to be reported`. I'm going to try closing the PR and re-opening it.

**Review by parthea — APPROVED** [review]

**Comment by parthea** [comment]
I'm going to try adding python `3.11.0-beta.3` to the tests to see if I can get the coverage check to pass

**Comment by brandonjbjelland** [comment]
👋  hey folks - where does this PR stand? It looks like the **exact** fix I'm needing for a build issue I'm currently hitting ([relevant lines](https://github.com/googleapis/proto-plus-python/pull/329/files#diff-11135ba0cd87468dca27e3a55d229e49647ff84f55d04301239ead576bd25743L61-R65)). Using both `3.10-slim` and `3.11-slim` base images that install `google.cloud.monitoring_v3`, my builds have started to fail with:

```
Step #2 - "install_dependencies_and_run_pytest": /usr/local/lib/python3.11/site-packages/proto/enums.py:61: in __new__
Step #2 - "install_dependencies_and_run_pytest":     idx = attrs._member_names.index(pb_options)
Step #2 - "install_dependencies_and_run_pytest": E   AttributeError: 'dict' object has no attribute 'index'
```

3.11 was released about a week ago - it may be worth trying that python build in the CI pipeline here.

**Comment by eremeevfd** [comment]
It would be great to have this compatibility fix merged :)

**Comment by parthea** [comment]
The github actions tests appear to be stuck. Closing and re-opening the PR didn't help. I'll try again tomorrow.

**Pull request merged**